reference/7.1/Microsoft.PowerShell.Core/About/about_Functions_Advanced_Argument_Completion.md

```diff
@@ -0,0 +1,364 @@
+    [String[]]
+    $Detail
+)
+```
+
+In the following example, the value of the variable `$flavor` must be either
+**Chocolate**, **Strawberry**, or **Vanilla**. The `ValidateSet` attribute can
+be used on any variable, not just parameters.
+
+```powershell
+[ValidateSet('Chocolate', 'Strawberry', 'Vanilla')]
+[String]$flavor = 'Strawberry'
+```
+
+The validation occurs whenever that variable is assigned even within the
+script. For example, the following results in an error at runtime:
+
+```powershell
+Param(
+    [ValidateSet('hello', 'world')]
+    [String]$Message
+)
+
+$Message = 'bye'
+```
+
+Beginning in PowerShell 7.0, tab expansion was added for the values of
+`ValidateSet`. For example, if you were typing the following variable
+definition:
+
+```
+[ValidateSet('Chocolate', 'Strawberry', 'Vanilla')][String]$flavor = 'S<Tab>
+```
+
+When you hit the <kbd>Tab</kbd> key, you would get the following result:
+
+```
+[ValidateSet('Chocolate', 'Strawberry', 'Vanilla')][String]$flavor = 'Strawberry'
+```
+
+For more information about tab expansion, see
+[about_Tab_Expansion](about_Tab_Expansion.md).
+
+### Dynamic ValidateSet values using classes
+
+You can use a **Class** to dynamically generate the values for **ValidateSet**
+at runtime. In the following example, the valid values for the variable
+`$Sound` are generated via a **Class** named **SoundNames** that checks three
+filesystem paths for available sound files:
+
+```powershell
+Class SoundNames : System.Management.Automation.IValidateSetValuesGenerator {
+    [String[]] GetValidValues() {
+        $SoundPaths = '/System/Library/Sounds/',
+                      '/Library/Sounds',
+                      '~/Library/Sounds'
+        $SoundNames = ForEach ($SoundPath in $SoundPaths) {
+            If (Test-Path $SoundPath) {
+                (Get-ChildItem $SoundPath).BaseName
+            }
+        }
+        return [String[]] $SoundNames
+    }
+}
+```
+
+The `[SoundNames]` class is then implemented as a dynamic **ValidateSet** value
+as follows:
+
+```powershell
+Param(
+    [ValidateSet([SoundNames])]
+    [String]$Sound
+)
+```
+
+> [!NOTE]
+> The `IValidateSetValuesGenerator` class was introduced in PowerShell 6.0.
+
+## ArgumentCompleter attribute
+
+The **ArgumentCompleter** attribute allows you to add tab completion values to
+a specific parameter. An **ArgumentCompleter** attribute must be defined for
+each parameter that needs tab completion. Similar to **DynamicParameters**, the
+available values are calculated at runtime when the user presses <kbd>Tab</kbd>
+after the parameter name.
+
+To add an **ArgumentCompleter** attribute, you need to define a script block
+that determines the values. The script block must take the following
+parameters in the order specified below. The parameter's names don't matter as
+the values are provided positionally.
+
+The syntax is as follows:
+
+```powershell
+Param(
+    [Parameter(Mandatory)]
+    [ArgumentCompleter( {
```

Comment:
Your examples both show the script as in-line, this is fine for small scripts, but for something more complex it should be a separate function, see https://github.com/PowerShell/PowerShell/blob/658837323599ab1c7a81fe66fcd43f7420e4402b/test/powershell/Language/Parser/ExtensibleCompletion.Tests.ps1#L132-L176 as an example

Reply:
Fixed